crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -288,290 +353,268 @@ impl AppHandlerExecute for Ics20Transfer {
+}
 
-    if is_source {
-        // the asset being transferred in is an asset that originated from astria
-        // subtract balance from escrow account and transfer to user
-
-        // strip the prefix from the denom, as we're back on the source chain
-        // note: if this is a refund, this is a no-op.
-        if !is_refund {
-            denom_trace.pop_trace_segment().context(
-                "there must be a source segment because above it was checked if the denom trace \
-                 contains a segment",
-            )?;
-        }
-        let escrow_channel = if is_refund {
-            source_channel
-        } else {
-            dest_channel
-        };
+#[instrument(skip_all)]
+async fn refund_tokens<S: StateWrite>(mut state: S, packet: &Packet) -> Result<()> {
+    let packet_data: FungibleTokenPacketData = serde_json::from_slice(&packet.data)
+        .context("failed to deserialize fungible token packet data")?;
 
-        let escrow_balance = state
-            .get_ibc_channel_balance(escrow_channel, &denom_trace)
-            .await
-            .context("failed to get IBC channel balance in execute_ics20_transfer")?;
+    let amount: u128 = packet_data
+        .amount
+        .parse()
+        .context("failed to parse packet data amount to u128")?;
 
-        state
-            .put_ibc_channel_balance(
-                escrow_channel,
-                &denom_trace,
-                escrow_balance.checked_sub(packet_amount).ok_or(anyhow!(
-                    "insufficient balance in escrow account to transfer tokens"
-                ))?,
+    // Since we are refunding tokens, packet_data.sender is an address on Astria:
+    // the packet was not commited on the counter party chain but returned as-is.
+    let receiver = parse_sender(&state, &packet_data.sender)
+        .await
+        .with_context(|| {
+            format!(
+                "failed parsing packet.sender `{}` as the return address",
+                packet_data.sender
+            )
+        })?;
+
+    let asset = parse_asset(&state, &packet_data.denom)
+        .await
+        .with_context(|| format!("failed parsing packet.asset `{}`", packet_data.denom))?;
+
+    if let Ok(memo) = serde_json::from_str::<Ics20WithdrawalFromRollup>(&packet_data.memo) {
+        info!(
+                memo.rollup_return_address,
+                memo_name = astria_core::generated::protocol::memos::v1alpha1::Ics20WithdrawalFromRollup::full_name(),
+                "was able to parse memo; refunding to rollup",
+            );
+        refund_tokens_to_rollup(&mut state, receiver, asset, amount, memo)
+            .await
+            .context("failed to refund the rollup")?;
+    } else {
+        info!(
+            recipient = packet_data.sender,
+            "memo in fungible token packet was unknown; taking packet sender as recipient address \
+             on sequencer",
+        );
+        let recipient = packet_data.sender.parse().with_context(|| {
```

Comment:
can you use the parsed `receiver` from above?

Reply:
Yes, thank you for noticing! I overlooked that bit. Done in 522f79bf